app/src/main/java/com/andrew/tasky/agenda/data/event/EventRepositoryImpl.kt

```diff
@@ -195,44 +201,56 @@ class EventRepositoryImpl @Inject constructor(
-        var photosDeleted = 0
-        val compressedEvent = event.copy(
-            photos = event.photos.mapNotNull { eventPhoto ->
-                if (eventPhoto is LocalPhoto) {
-                    if (eventPhoto.savedInternally) {
-                        val byteArray = imageStorage.getByteArray(eventPhoto.key)
-                        eventPhoto.copy(byteArray = byteArray)
-                    } else {
-                        val byteArray = eventPhoto.bitmap?.let { bitmap ->
-                            bitmapConverter.bitmapToCompressByteArray(
-                                bitmap, MAX_PHOTO_SIZE_IN_BYTES
-                            )
-                        }
-                        if (byteArray != null) {
-                            eventPhoto.copy(byteArray = byteArray)
-                        } else {
-                            photosDeleted++
-                            null
-                        }
-                    }
-                } else {
-                    eventPhoto
+        return withContext(Dispatchers.Default) {
+            var photosDeleted = 0
+            val compressedEvent = event.copy(
+                photos = supervisorScope {
+                    event.photos
+                        .chunked(MAX_PARALLEL_IMAGE_COMPRESS_COUNT)
+                        .map { eventPhotoChunk ->
+                            eventPhotoChunk.map { eventPhoto ->
+                                async {
+                                    if (eventPhoto is LocalPhoto) {
+                                        if (eventPhoto.savedInternally) {
+                                            val byteArray =
+                                                imageStorage.getByteArray(eventPhoto.key)
+                                            eventPhoto.copy(byteArray = byteArray)
+                                        } else {
+                                            val byteArray = eventPhoto.bitmap?.let { bitmap ->
+                                                bitmapConverter.bitmapToCompressByteArray(
+                                                    bitmap, MAX_PHOTO_SIZE_IN_BYTES
+                                                )
+                                            }
+                                            if (byteArray != null) {
+                                                eventPhoto.copy(byteArray = byteArray)
+                                            } else {
+                                                photosDeleted++
+                                                null
+                                            }
+                                        }
+                                    } else {
+                                        eventPhoto
+                                    }
+                                }
+                            }.mapNotNull { it.await() }
+                        }.flatten()
```

Comment:
Is it best to put this in a class inside the events data package? Can you elaborate what belongs in the repo class and what doesn't?

Reply:
The repository takes in multiple data sources (such as API, local DB, image cache), combines them and then decides what data to expose to the domain/presentation layer.
Here in this case the compression logic is quite complex and just bloats the repository. In such cases I like to put that in a separate class/file so that this logic can be tested more easily (in your case you'd need an instance of your repo which needs an API, a DB and more - all of which are things the compression logic doesn't need). You can also make it part of your bitmap converter, it doesn't necessarily need to be in an otherwise empty class.

I know it's a bit difficult to get a feeling for that and there's also not always a black & white answer, but hope this clarified it a bit